### PR TITLE
Replace tags in attribute lists and dicts

### PR DESCRIPTION
**Pull request by korsbakken**
This PR (hopefully) solves issue #342 . It modifies `Code.replace_tag` to descend into attribute lists and dicts, replaing each string in list elements of dict values that contain a tag. Adds a test and test data that contains a code with tags in the `components` attribute.

### CODE REVIEW

File: tests/data/tagged_codelist/foo_attr_list.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+- Emissions|Species:
+    # the "Sector" guards against feature regression (conflict with "Sector" tag)
+    definition: Emissions of Species
+    unit: tSpecies/yr
+
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/foo_attr_list.yaml:5]: This feature regression is already tested in `foo.yaml` so suggest to remove here for simplicity.
```suggestion
```

File: tests/data/tagged_codelist/tag_species.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+- Species:
+  - CO2:
+     definition: CO2
+     unit: tCO2
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/tag_species.yaml:4]: For consistency with the species in [common-definitions](https://github.com/IAMconsortium/common-definitions/blob/e009dfe10b3eb2c8440b114e44ee3e929c0cabae/definitions/variable/emissions/tag_level-1-species.yaml#L4).

Also, if "definition" would not be specified here, the nomenclature package would replace the tag in "definition" in the code item by the name (i.e., also "CO2"), so replacing it with a longer string helps to make sure that it uses the right replacement (the replacement from the "definition", not the item name.)
```suggestion
     definition: carbon dioxide (CO2)
     unit: Mt CO2
```

File: tests/data/tagged_codelist/tag_species.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  - CH4:
+      definition: CH4
+      unit: tCH4
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/tag_species.yaml:7]:
```suggestion
        description: methane (CH4)
        unit: Mt CH4/yr

```

File: tests/test_codelist.py [hunk 4]
```diff
@@ -110,6 +110,45 @@ def test_tagged_codelist():
...
+        "variable", TEST_DATA_DIR / "tagged_codelist"
+    )
+    # The test should test that the tags in the definitions in the
+    # tagged_codelist/foo_attr_list_dict.yaml file are expanded correctly.
+
+    exp = {
+        "Emissions|Species": {
+            "description": "Emissions of Species",
+            "unit": "tSpecies/yr",
+        },
```
Comment (danielhuppmann) [inline tests/test_codelist.py:125]:
```suggestion
```

File: tests/test_codelist.py [hunk 5]
```diff
@@ -110,6 +110,45 @@ def test_tagged_codelist():
...
+        "Emissions|CO2": {
+            "description": "Total emissions of CO2",
+            "unit": "tCO2/yr",
```
Comment (danielhuppmann) [inline tests/test_codelist.py:128]:
```suggestion
            "description": "Total emissions of carbon dioxide (CO2)",
            "unit": "Mt CO2/yr",

```

File: tests/test_codelist.py [hunk 6]
```diff
@@ -110,6 +110,45 @@ def test_tagged_codelist():
...
+            "check_aggregate": True,
+            "components": {
+                "By source": ["Emissions|CO2|Fossil", "Emissions|CO2|Renewables"],
+                "By sector": ["Emissions|CO2|Energy", "Emissions|CO2|Industry"],
+            },
+        },
+        "Emissions|CH4": {
+            "description": "Total emissions of CH4",
+            "unit": "tCH4/yr",
```
Comment (danielhuppmann) [inline tests/test_codelist.py:137]:
```suggestion
            "description": "Total emissions of methane (CH4)",
            "unit": "Mt CH4/yr",
```

File: nomenclature/code.py [hunk 7]
```diff
@@ -112,21 +112,28 @@ def replace_tag(self, tag: str, target: "Code") -> "Code":
...
-                    )
-                # otherwise insert the name
+                if _attr in target.flattened_dict:
+                    return _value.replace("{" + tag + "}", getattr(target, _attr))
+                # otherwise return the name
                 else:
-                    mapping[attr] = value.replace("{" + tag + "}", target.name)
-            # otherwise append as is
+                    return _value.replace("{" + tag + "}", getattr(target, "name"))
+            # If the attribute is a list, descend into it and replace
```
Comment (danielhuppmann) [inline nomenclature/code.py:124]:
```suggestion
            # if the attribute is a list, iterate over the items and replace tags
```

File: nomenclature/code.py [hunk 8]
```diff
@@ -112,21 +112,28 @@ def replace_tag(self, tag: str, target: "Code") -> "Code":
...
+            elif isinstance(_value, list):
+                return [_replace_or_recurse(_attr, _v) for _v in _value]
+            # If the attribute is a dict, descend into it and replace the values
```
Comment (danielhuppmann) [inline nomenclature/code.py:127]:
```suggestion
            # if the attribute is a mapping, iterate over the items and replace tags
            # in the values (not the keys)
```

**Review by danielhuppmann — COMMENTED** [review]
Thanks @korsbakken, very nice - just a few suggestions for consistency...

**Comment by korsbakken** [comment]
> Thanks @korsbakken, very nice - just a few suggestions for consistency...

Thanks @danielhuppmann, those suggestions make good sense, I have added and committed them to the PR now.

**Comment by korsbakken** [comment]
The suggested changes appear to cause tests to fail, not clear why. It introduces an unintentional blank line in the middle of a definition, and an inconsistency in the CO2 and CH4 unit definitions, but new less obvious failures appear after fixing that.

File: tests/data/tagged_codelist/foo_attr_list.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+- Emissions|{Species}:
+    definition: Total emissions of {Species}
+    unit: "{Species}/yr"
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/foo_attr_list.yaml:3]:
```suggestion
    unit: "{Species}"
```

File: tests/data/tagged_codelist/tag_species.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+- Species:
+  - CO2:
+     definition: carbon dioxide (CO2)
+     unit: Mt CO2
```
Comment (danielhuppmann) [inline tests/data/tagged_codelist/tag_species.yaml:4]:
```suggestion
        definition: carbon dioxide (CO2)
        unit: Mt CO2/yr
```

**Review by danielhuppmann — COMMENTED** [review]
Ha, the "/yr" was specified both in the code itself and the tag...

File: tests/data/tagged_codelist/tag_species.yaml (hunk 10 above)
Comment (korsbakken) [reply tests/data/tagged_codelist/tag_species.yaml]: Thanks. There was also inconsistent indentation in tag_species.yaml that was causing errors. I had already made fixes locally. I have committed and pushed those now, so I haven't committed your suggestions.

My commit is slightly different from yours. I kept `unit: {Species}/yr` in the `Emissions|{Species}` tagged variable, and instead removed `/yr` from both unit attributes in tag_species.yaml. The species tag could in principle also be used for variables representing cumulative emissions, emission intensity, etc. Those could all have units with `Mt {Species}` in the denominator, but would need different or no units in the denominator.

It doesn't matter here of course, so I'm happy to move where `/yr` goes if you have a strong preference.

**Review by danielhuppmann — APPROVED** [review]
Looks good, thanks!

File: tests/data/tagged_codelist/tag_species.yaml (hunk 10 above)
Comment (danielhuppmann) [reply tests/data/tagged_codelist/tag_species.yaml]: Good point, thanks.

**Pull request merged**